<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Unsupervised multiomics data integration</title>
    <meta charset="utf-8" />
    <meta name="author" content="Amrit Singh, PhD Department of Anesthesiology, Pharmacology and Therapeutics, UBC  Centre for Heart Lung Innovation" />
    <script src="libs/header-attrs-2.14/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/middlebury-fonts.css" rel="stylesheet" />
    <link rel="stylesheet" href="theme.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, my-title, title-slide

.title[
# Unsupervised multiomics data integration
]
.author[
### Amrit Singh, PhD<br><span style="font-size: 65%;">Department of Anesthesiology, Pharmacology and Therapeutics, UBC<br> Centre for Heart Lung Innovation</span>
]
.date[
### August 08, 2022, 2022 | 09:00-11:00<br><br><svg aria-hidden="true" role="img" viewBox="0 0 640 512" style="height:1em;width:1.25em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:currentColor;overflow:visible;position:relative;"><path d="M172.5 131.1C228.1 75.51 320.5 75.51 376.1 131.1C426.1 181.1 433.5 260.8 392.4 318.3L391.3 319.9C381 334.2 361 337.6 346.7 327.3C332.3 317 328.9 297 339.2 282.7L340.3 281.1C363.2 249 359.6 205.1 331.7 177.2C300.3 145.8 249.2 145.8 217.7 177.2L105.5 289.5C73.99 320.1 73.99 372 105.5 403.5C133.3 431.4 177.3 435 209.3 412.1L210.9 410.1C225.3 400.7 245.3 404 255.5 418.4C265.8 432.8 262.5 452.8 248.1 463.1L246.5 464.2C188.1 505.3 110.2 498.7 60.21 448.8C3.741 392.3 3.741 300.7 60.21 244.3L172.5 131.1zM467.5 380C411 436.5 319.5 436.5 263 380C213 330 206.5 251.2 247.6 193.7L248.7 192.1C258.1 177.8 278.1 174.4 293.3 184.7C307.7 194.1 311.1 214.1 300.8 229.3L299.7 230.9C276.8 262.1 280.4 306.9 308.3 334.8C339.7 366.2 390.8 366.2 422.3 334.8L534.5 222.5C566 191 566 139.1 534.5 108.5C506.7 80.63 462.7 76.99 430.7 99.9L429.1 101C414.7 111.3 394.7 107.1 384.5 93.58C374.2 79.2 377.5 59.21 391.9 48.94L393.5 47.82C451 6.731 529.8 13.25 579.8 63.24C636.3 119.7 636.3 211.3 579.8 267.7L467.5 380z"/></svg> <a href="https://cbl-hli.github.io/">lab</a><br><svg aria-hidden="true" role="img" viewBox="0 0 496 512" style="height:1em;width:0.97em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:currentColor;overflow:visible;position:relative;"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3 .3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5 .3-6.2 2.3zm44.2-1.7c-2.9 .7-4.9 2.6-4.6 4.9 .3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3 .7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3 .3 2.9 2.3 3.9 1.6 1 3.6 .7 4.3-.7 .7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3 .7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3 .7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> <a href="https://github.com/Phillip-a-richmond/PrecisionHealthVirtualEnvironment/tree/main/Workshops">code</a><br><svg aria-hidden="true" role="img" viewBox="0 0 512 512" style="height:1em;width:1em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:currentColor;overflow:visible;position:relative;"><path d="M459.4 151.7c.325 4.548 .325 9.097 .325 13.65 0 138.7-105.6 298.6-298.6 298.6-59.45 0-114.7-17.22-161.1-47.11 8.447 .974 16.57 1.299 25.34 1.299 49.06 0 94.21-16.57 130.3-44.83-46.13-.975-84.79-31.19-98.11-72.77 6.498 .974 12.99 1.624 19.82 1.624 9.421 0 18.84-1.3 27.61-3.573-48.08-9.747-84.14-51.98-84.14-102.1v-1.299c13.97 7.797 30.21 12.67 47.43 13.32-28.26-18.84-46.78-51.01-46.78-87.39 0-19.49 5.197-37.36 14.29-52.95 51.65 63.67 129.3 105.3 216.4 109.8-1.624-7.797-2.599-15.92-2.599-24.04 0-57.83 46.78-104.9 104.9-104.9 30.21 0 57.5 12.67 76.67 33.14 23.72-4.548 46.46-13.32 66.6-25.34-7.798 24.37-24.37 44.83-46.13 57.83 21.12-2.273 41.58-8.122 60.43-16.24-14.29 20.79-32.16 39.31-52.63 54.25z"/></svg> <a href="https://twitter.com/asingh_22g">asingh_22g</a>
]

---





background-image: url(img/la.svg)

---
background-image: url(img/cc.svg)

---

class: middle

# Learning outcomes

### 1. Contrast between unsupervised and supervised learning
### 2. Develop an intuition for how multiomics methods are trying to explain the underlying biology
### 3. Give examples of methods that integrate multiomics data and what each is trying to acheive
### 4. Apply a few unsupervised integrative methods to real world data

---

*"In supervised learning, the goal is to predict the value of an outcome measure based on a number of input measures;
in unsupervised learning, there is no outcome measure, and the goal is to
describe the associations and patterns among a set of input measures."* [Elements of statistical learning, 2008 page 3](https://hastie.su.domains/Papers/ESLII.pdf)

.pull-left[

### Unsupervised (clustering)
.center[&lt;img src ="img/kmeans.gif" alt="kmeans", width="80%"/&gt;]
[Chire 2017](https://commons.wikimedia.org/wiki/File:K-means_convergence.gif)
]

.pull-right[

### Supervised (regression/classification)
.center[&lt;img src ="img/decision_boundary.gif" alt="decision_boundary", width="80%"/&gt;]
[Tensorflow playground](https://playground.tensorflow.org/)
]


---

# Unsupervised multiomics

.center[&lt;img src ="img/overview_unsupervised.png" alt="Overview of Unsupervised Multiomics", width="80%"/&gt;]

---

# Terminology review


| Term      | Synonym | Description
| ----------- | ----------- |
| multiomics      |    multiview, multidomain, mulitmodal    | confirmation of findings using the same samples, (e.g. measuring the same protein using different technologies in the same set of samples)
| validation   |  | confirmation of findings using samples from the original discovery cohort
| dataset | block, table, data matrix | a 2D matrix comprises of observations (*n*) and variables (*p*)
|latent variable | component, embedding, hidden variable | low-dimensional variable that contains information about the high dimensional dataset

---

.center[&lt;img src ="img/multiview.png" alt="Multiview tables", width="110%"/&gt;]
.right[&lt;img style="margin-top: -20rem" src ="img/susan_holmes.png" alt="Susan Holmes", width="20%"/&gt;]
[Sankaran K and Holmes SP. Front Genet. 2019 Aug 28;10:627](https://pubmed.ncbi.nlm.nih.gov/31555316/)

---

# Concatenated PCA 

$$\Huge X = \left[ X^{1} | ... | X^{J} \right] $$


```r
pca_res &lt;- breast.TCGA$data.train[1:3] %&gt;% 
  do.call(cbind, .) %&gt;% 
  prcomp(x = ., rank. = 2, center = TRUE, scale. = TRUE)
```

.pull-left[
&lt;img src="slides_files/figure-html/unnamed-chunk-3-1.png" width="60%" /&gt;
]

.pull-right[
&lt;img src="slides_files/figure-html/unnamed-chunk-4-1.png" width="70%" /&gt;
]
---

# Limitation of Concatenated PCA

- provides relationship between variables but not sets of variables
- datasets with more variables can dominate the resulting PCs; the mirna, mrna, protein datasets have 184, 200, 142 variables respectively.

.pull-left[

```r
p &lt;- lapply(breast.TCGA$data.train[1:3], function(X){
  result &lt;- prcomp(x = X, rank. = 4, center = TRUE, scale. = TRUE)
  propvar &lt;- 100*(result$sdev^2)/sum(result$sdev^2)
  propvar[1:4]
}) %&gt;% 
  do.call(cbind, .) %&gt;% 
  as.data.frame() %&gt;% 
  mutate(ncomp = 1:4) %&gt;% 
  gather(dataset, propvar, -ncomp) %&gt;% 
  ggplot(aes(x = ncomp, y = propvar, fill = dataset)) +
  geom_bar(stat="identity", position=position_dodge()) +
  ylab("Proportion of variation explaind (%)")
```
]

.pull-right[
&lt;img src="slides_files/figure-html/unnamed-chunk-6-1.png" width="80%" /&gt;
]

---

# Weighted PCA (Multiple Factor Analysis, MFA)

$$\Huge X = \left[ \frac{X^{1}}{\lambda_1(X^{1})} | ... | \frac{X^{J}}{\lambda_1(X^{J})} \right] $$


```r
res &lt;- FactoMineR::MFA(as.data.frame(do.call(cbind, breast.TCGA$data.train[1:3])), 
           group=sapply(breast.TCGA$data.train[1:3], ncol), type=rep("s",3),
    ncp=5, name.group=names(breast.TCGA$data.train[1:3]), graph=FALSE)
```
[Guide to MFA interpretation](https://www.youtube.com/watch?v=OVtNX6Or1FI&amp;ab_channel=Fran%C3%A7oisHusson)

???

Groups representation: correlation between each group and the principal component. mrna has stronger correlation with the first PC whereas protein has a stronger correlation with the second PC. mrna and mirna groups have are correlated with each other and induce a similar structure

---

# MFA

.pull-left[

```r
plot(res,choix="var", select = "cos2 0.6")
```

![](slides_files/figure-html/unnamed-chunk-8-1.png)&lt;!-- --&gt;

]

.pull-right[

```r
plot(res,choix="ind",partial="all")
```

```
## Warning: ggrepel: 122 unlabeled data points (too many overlaps). Consider
## increasing max.overlaps
```

![](slides_files/figure-html/unnamed-chunk-9-1.png)&lt;!-- --&gt;

```r
# res$global.pca$ind$coord[,1:2] %&gt;% 
#   as.data.frame() %&gt;% 
#   mutate(subtype=breast.TCGA$data.train$subtype) %&gt;% 
#   ggplot(aes(x=Dim.1, y=Dim.2, color=subtype)) +
#   geom_point() +
#   theme_classic()
```
]

---

# Multi-Omics Factor Analysis (MOFA)
- PCA for multiple datasets; feature selection for associated latent variables

.center[&lt;img src ="img/mofa.png" alt="MOFA", width="80%"/&gt;]

[MOFA](https://biofam.github.io/MOFA2/)
[Genome Biology volume 21, Article number: 111 (2020) ](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02015-1)

---

# MOFA

.pull-left[

```r
library(MOFA2)

MOFAobject &lt;- create_mofa(lapply(breast.TCGA$data.train[1:3], t))

data_opts &lt;- get_default_data_options(MOFAobject)
model_opts &lt;- get_default_model_options(MOFAobject)
train_opts &lt;- get_default_training_options(MOFAobject)

MOFAobject &lt;- prepare_mofa(
  object = MOFAobject,
  data_options = data_opts,
  model_options = model_opts,
  training_options = train_opts
)
```


```r
p &lt;- plot_data_overview(MOFAobject)
```


]

.pull-right[
![](slides_files/figure-html/unnamed-chunk-12-1.png)&lt;!-- --&gt;
]


---

# MOFA

.pull-left[

```r
outfile = file.path(getwd(),"model.hdf5")
model &lt;- run_mofa(MOFAobject, outfile, use_basilisk = FALSE)
samples_metadata(model) &lt;- data.frame(
  sample = samples_names(model)[[1]],
  subtype = breast.TCGA$data.train$subtype
)

p &lt;- plot_variance_explained(model)
```
]

.pull-right[
![](slides_files/figure-html/unnamed-chunk-14-1.png)&lt;!-- --&gt;
]

---

# MOFA

.pull-left[

```r
plot_factors(model, 
  factors = 1:3,
  color_by = "subtype"
)
```

&lt;img src="slides_files/figure-html/unnamed-chunk-15-1.png" width="70%" /&gt;
]

.pull-right[

```r
plot_top_weights(model,
  view = "mrna",
  factors = 1,
  nfeatures = 10
)
```

&lt;img src="slides_files/figure-html/unnamed-chunk-16-1.png" width="70%" /&gt;

]

---

# MOFA + tSNE

.pull-left[

```r
set.seed(42)
# model &lt;- run_umap(model)
model &lt;- run_tsne(model)

p &lt;- plot_dimred(model,
  method = "TSNE",  # method can be either "TSNE" or "UMAP"
  color_by = "subtype"
)
```

```
## Warning: `guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; =
## "none")` instead.
```
]

.pull-right[
![](slides_files/figure-html/unnamed-chunk-18-1.png)&lt;!-- --&gt;

]

---

# Partial Least Squares (PLS)

- unlike PCA which maximizes variance, PLS maximize covariance between two data matrices
- doesnt work with more than 2 data matrices



.pull-left[

```r
cim(cor(pls_res$variates$X, pls_res$variates$Y), margins = c(10, 10), 
    xlab = "mRNA", ylab = "miRNA")
```

![](slides_files/figure-html/unnamed-chunk-20-1.png)&lt;!-- --&gt;

]

.pull-right[

```r
plotIndiv(pls_res, rep.space = "XY-variate", group = breast.TCGA$data.train$subtype,
          ellipse = TRUE)
```

![](slides_files/figure-html/unnamed-chunk-21-1.png)&lt;!-- --&gt;

]

- similar to the canonical correlation analysis used in [Seurat, RunCCA()](https://satijalab.org/seurat/reference/runcca)

---

# PLS

.pull-left[

```r
plotVar(pls_res, cutoff = 0.6)
```

&lt;img src="slides_files/figure-html/unnamed-chunk-22-1.png" width="90%" /&gt;

]

.pull-right[

```r
plotLoadings(pls_res)
```

&lt;img src="slides_files/figure-html/unnamed-chunk-23-1.png" width="90%" /&gt;

]

---

# Similarity Network Fusion (SNF)

.center[&lt;img src ="img/snf.png" alt="SNF", width="50%"/&gt;]

---

# ?SNF

.pull-left[

```r
library(SNFtool)

## First, set all the parameters:
K = 20;		# number of neighbors, usually (10~30)
alpha = 0.5;  	# hyperparameter, usually (0.3~0.8)
T = 20; 	# Number of Iterations, usually (10~20)

std_norm &lt;- breast.TCGA$data.train[1:3]
std_norm$mrna &lt;- standardNormalization(std_norm$mrna)
std_norm$mirna &lt;- standardNormalization(std_norm$mirna)

W &lt;- lapply(std_norm, function(i){
  # distance matrix
  dist_matrix &lt;- (dist2(as.matrix(i),as.matrix(i)))^(1/2)
  # affinity matrix calculation
  affinityMatrix(dist_matrix, K, alpha)
})

W = SNF(W, K, T)


C = 3
labels = spectralClustering(W, C)
```
]

.pull-right[

```r
displayClusters(W, labels)
```

![](slides_files/figure-html/unnamed-chunk-25-1.png)&lt;!-- --&gt;

]

---

# Unpaired multiomics data integration (GLUE)

.center[&lt;img src ="img/glue.png" alt="GLUE", width="60%"/&gt;]
.right[&lt;img style="margin-top: -20rem" src ="img/ge_gao.jpeg" alt="Susan Holmes", width="15%"/&gt;]
[Cao Z and Gao G. Nat Biotechnol. 2022 May 2.](https://pubmed.ncbi.nlm.nih.gov/35501393/) | [GLUE code](https://github.com/gao-lab/GLUE)

???

- each omics layer is fit with a variational autoencoder
- knowledge-based graph models the cross-layer regulatory interactions for linking layer-specific feature spaces; e.g. for scRNAseq and scATACseq, vertices (genes and accessible chromatin regions) and edges (connection accesible region and downstream gene)
- u: cell latent variable; v: feature latent variable
- adversarial multimodal alignment strategy; predicts omics layer based on their embeddings.

---

# timeOmics

.center[&lt;img src ="img/timeomics.png" alt="Time Omics", width="60%"/&gt;]

[Bodein et al., Front Genet. 2019 Nov 7;10:963.](https://pubmed.ncbi.nlm.nih.gov/31803221/)

---

## Single cell data integration

.center[&lt;img src ="img/scDataIntegration.png" alt="single cell data integration", width="80%"/&gt;]
[Nat Methods. 2022 Jan;19(1):41-50.](https://pubmed.ncbi.nlm.nih.gov/34949812/) |
[Explanation of metrics](https://openproblems.bio/neurips_docs/about_tasks/task3_joint_embedding/)

---

# Assessing overfitting and consistency

.center[&lt;img src ="img/cv.png" alt="Cross validation", width="60%"/&gt;]

[Brief Bioinform. 2020 Jul 15;21(4):1277-1284](https://pubmed.ncbi.nlm.nih.gov/31281919/)

---

.center[&lt;img src ="img/book_chapter.png" alt="Book chapter", width="100%"/&gt;]
[Singh A et al., Handbook of Biomarkers and Precision Medicine CRC Press 2019:596](https://www.taylorfrancis.com/books/edit/10.1201/9780429202872/handbook-biomarkers-precision-medicine-claudio-carini-mark-fidock-alain-van-gool)


---

class: middle, center

background-image: url(img/bkg.svg)

# THANK YOU!

August 08, 2022, 2022 | 09:00-11:00&lt;br&gt;&lt;br&gt;<svg aria-hidden="true" role="img" viewBox="0 0 640 512" style="height:1em;width:1.25em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:currentColor;overflow:visible;position:relative;"><path d="M172.5 131.1C228.1 75.51 320.5 75.51 376.1 131.1C426.1 181.1 433.5 260.8 392.4 318.3L391.3 319.9C381 334.2 361 337.6 346.7 327.3C332.3 317 328.9 297 339.2 282.7L340.3 281.1C363.2 249 359.6 205.1 331.7 177.2C300.3 145.8 249.2 145.8 217.7 177.2L105.5 289.5C73.99 320.1 73.99 372 105.5 403.5C133.3 431.4 177.3 435 209.3 412.1L210.9 410.1C225.3 400.7 245.3 404 255.5 418.4C265.8 432.8 262.5 452.8 248.1 463.1L246.5 464.2C188.1 505.3 110.2 498.7 60.21 448.8C3.741 392.3 3.741 300.7 60.21 244.3L172.5 131.1zM467.5 380C411 436.5 319.5 436.5 263 380C213 330 206.5 251.2 247.6 193.7L248.7 192.1C258.1 177.8 278.1 174.4 293.3 184.7C307.7 194.1 311.1 214.1 300.8 229.3L299.7 230.9C276.8 262.1 280.4 306.9 308.3 334.8C339.7 366.2 390.8 366.2 422.3 334.8L534.5 222.5C566 191 566 139.1 534.5 108.5C506.7 80.63 462.7 76.99 430.7 99.9L429.1 101C414.7 111.3 394.7 107.1 384.5 93.58C374.2 79.2 377.5 59.21 391.9 48.94L393.5 47.82C451 6.731 529.8 13.25 579.8 63.24C636.3 119.7 636.3 211.3 579.8 267.7L467.5 380z"/></svg> [lab](https://cbl-hli.github.io/)&lt;br&gt;<svg aria-hidden="true" role="img" viewBox="0 0 496 512" style="height:1em;width:0.97em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:currentColor;overflow:visible;position:relative;"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3 .3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5 .3-6.2 2.3zm44.2-1.7c-2.9 .7-4.9 2.6-4.6 4.9 .3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3 .7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3 .3 2.9 2.3 3.9 1.6 1 3.6 .7 4.3-.7 .7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3 .7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3 .7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> [code](https://github.com/Phillip-a-richmond/PrecisionHealthVirtualEnvironment/tree/main/Workshops)&lt;br&gt;<svg aria-hidden="true" role="img" viewBox="0 0 512 512" style="height:1em;width:1em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:currentColor;overflow:visible;position:relative;"><path d="M459.4 151.7c.325 4.548 .325 9.097 .325 13.65 0 138.7-105.6 298.6-298.6 298.6-59.45 0-114.7-17.22-161.1-47.11 8.447 .974 16.57 1.299 25.34 1.299 49.06 0 94.21-16.57 130.3-44.83-46.13-.975-84.79-31.19-98.11-72.77 6.498 .974 12.99 1.624 19.82 1.624 9.421 0 18.84-1.3 27.61-3.573-48.08-9.747-84.14-51.98-84.14-102.1v-1.299c13.97 7.797 30.21 12.67 47.43 13.32-28.26-18.84-46.78-51.01-46.78-87.39 0-19.49 5.197-37.36 14.29-52.95 51.65 63.67 129.3 105.3 216.4 109.8-1.624-7.797-2.599-15.92-2.599-24.04 0-57.83 46.78-104.9 104.9-104.9 30.21 0 57.5 12.67 76.67 33.14 23.72-4.548 46.46-13.32 66.6-25.34-7.798 24.37-24.37 44.83-46.13 57.83 21.12-2.273 41.58-8.122 60.43-16.24-14.29 20.79-32.16 39.31-52.63 54.25z"/></svg> [asingh_22g](https://twitter.com/asingh_22g)
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
