<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Biomarker discovery and evaluation techniques</title>
    <meta charset="utf-8" />
    <meta name="author" content="Amrit Singh, PhD Department of Anesthesiology, Pharmacology and Therapeutics, UBC  Centre for Heart Lung Innovation" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/middlebury-fonts.css" rel="stylesheet" />
    <link rel="stylesheet" href="theme.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, my-title, title-slide

.title[
# Biomarker discovery and evaluation techniques
]
.author[
### Amrit Singh, PhD<br><span style="font-size: 65%;">Department of Anesthesiology, Pharmacology and Therapeutics, UBC<br> Centre for Heart Lung Innovation</span>
]
.date[
### August 05, 2022 | 12:00-14:00<br><br><svg aria-hidden="true" role="img" viewBox="0 0 512 512" style="height:1em;width:1em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:currentColor;overflow:visible;position:relative;"><path d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"/></svg> <a href="https://cbl-hli.github.io/">lab</a><br><svg aria-hidden="true" role="img" viewBox="0 0 496 512" style="height:1em;width:0.97em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:currentColor;overflow:visible;position:relative;"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> <a href="https://github.com/Phillip-a-richmond/PrecisionHealthVirtualEnvironment/tree/main/Workshops">code</a><br><svg aria-hidden="true" role="img" viewBox="0 0 512 512" style="height:1em;width:1em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:currentColor;overflow:visible;position:relative;"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> <a href="https://twitter.com/asingh_22g">asingh_22g</a>
]

---





background-image: url(img/la.svg)

---
background-image: url(img/cc.svg)

---

class: middle

# Learning outcomes

### 1. Differentiate between Git and GitHub and be able to setup the connection on sockeye
### 2. Use git to copy open-source code and test the code locally
### 3. Make changes and update codebase (including undo)
### 4. Implement the basics of Git Flow

---

# What is a biomarker?

*A defined characteristic that is measured as an indicator of normal biologial processes, pathogenic processes, or biological responses to an exposure or intervention, including therapeutic interventions.*

.pull-left[
#### Types
- molecular
- histologic
- radiographic
- physiologic
]

.pull-right[
#### Categories
- susceptibility/risk biomarker
- diagnostic biomarker
- monitoring biomarker
- prognostic biomarker
- predictive biomarker
- response biomarker
- safety biomarker
]

&lt;br&gt;

[FDA-NIH Biomarker Working Group. BEST (Biomarkers, EndpointS, and other Tools) Resource [Internet]. Silver Spring (MD): Food and Drug Administration (US); 2016-. Available from: https://www.ncbi.nlm.nih.gov/books/NBK326791/ Co-published by National Institutes of Health (US), Bethesda (MD).](https://www.ncbi.nlm.nih.gov/books/NBK326791/)
]

---

## Diagnostic biomarker: 


---

# Magic bullet biomarkers: do they exist?

.pull-left[
&lt;img src ="img/ntbnp.png" alt="NTpBNP", width="85%"/&gt;
]

.pull-right[
**N terminal probrain natriuretic peptide (NTproBNP)**
is a rule-out test for acute heart failure

serum NTpBNP &lt;= 300ng/L :
- sensitivity: 0.95 (0.93-0.96)
- negative predictive value: 0.98 (0.89-1)
- specificity: 0.43 (0.26-0.62)
- positive predictive value: 0.64 (0.57-0.73)

cardic imaging is required to confirm diagnosis if positive for heart failure.
]

[Roberts E et al., BMJ. 2015 Mar 4;350:h910.](https://pubmed.ncbi.nlm.nih.gov/24650281/)


---

# Why use a multimarker approach?

.center[&lt;img src ="img/characteristic_direction.png" alt="Characteristic Direction", width="55%"/&gt;]

[BMC Bioinformatics
. 2014 Mar 21;15:79. doi: 10.1186/1471-2105-15-79.](https://pubmed.ncbi.nlm.nih.gov/24650281/)


---

.center[&lt;img src ="img/multimarker.png" alt="Multimarker", width="100%"/&gt;]

[Singh A et al., AJRCCM 2018, 15:79-93.
](https://pubmed.ncbi.nlm.nih.gov/24650281/)
[Portt L. et al. Biochimica et Biophysica Acta 2011, 238-259]()


---


## Clinical utlity?

---


# Terminology review

| Term      | Synonym | Description
| ----------- | ----------- |
| Technical replication      |        | confirmation of findings using the same samples, (e.g. measuring the same protein using different technologies in the same set of samples)
| validation   |  | confirmation of findings using samples from the original discovery cohort
| biomarker panel | biomarker signature, multimarker panel | combination of multimarker panels



---

# How to identify biomarkers?

choose your process:
* differential expression
* machine learning methods
* combination of approaches

---

### Method 1
 - Step 1: Differential expression (50 gene panel)
 - Step 2: train machine learning model

.pull-left[
&lt;img src="slides_files/figure-html/unnamed-chunk-3-1.png" width="80%" /&gt;
]

.pull-right[
&lt;img src="slides_files/figure-html/unnamed-chunk-4-1.png" width="80%" /&gt;

]

&gt; ZNF552, KDM4B, C4orf34, LRIG1, PREX1, FUT8, CCNA2, TTC39A, STC2, SLC43A3, SEMA3C, ASPM, MEX3A, MED13L, DTWD2, ELP2, LMO4, E2F1, FAM63A, SLC19A2, FMNL2, MEGF9, ALCAM, EPHB3, CSRP2, MTL5, TANC2, CDK18, KIF13B, SNED1, HN1, OGFRL1, NTN4, SNORA8, YPEL2, NCAPG2, SLC5A6, HTRA1, LIMD2, ALDH6A1, SHROOM3, FRMD6, SIGIRR, IFITM2, RTN2, RIN3, C18orf1, ASF1B, PBX1, ZEB1

---

## Method 2
- using a machine learning model with variable selection (select 50 genes): sPLS-DA

.pull-left[
&lt;img src="slides_files/figure-html/unnamed-chunk-5-1.png" width="80%" /&gt;
]

.pull-right[
&lt;img src="slides_files/figure-html/unnamed-chunk-6-1.png" width="80%" /&gt;
]

&gt; ZNF552, KDM4B, PREX1, LRIG1, CCNA2, TTC39A, C4orf34, FUT8, ASPM, MEX3A, SLC43A3, SEMA3C, STC2, LMO4, E2F1, MED13L, FMNL2, DTWD2, SLC19A2, KIF13B, FAM63A, CSRP2, NTN4, MTL5, EPHB3, TP53INP2, CDK18, NDRG2, TRIM45, STAT5A, PLEKHA4, ZNF37B, OGFRL1, PLCD3, NES, ELP2, SDC1, STC2, PROM2, AMN1, PSIP1, S100A16, DSG2, DBP, TANC2, CD302, CPT1A, EGLN3, ALCAM, MXI1

---

class: center

## Overlap between biomarkers between the two approaches

![](slides_files/figure-html/unnamed-chunk-7-1.png)&lt;!-- --&gt;

---

# How to determine the performance of biomarkers?

| Metric      | Description |
| :---        |    :----   |
| Area under the receiver &lt;br&gt; operating characteristic curve (AUROC)     | &lt;ul&gt;&lt;li&gt; binary classification&lt;/li&gt; &lt;li&gt;true vs. false positive rate&lt;/li&gt;&lt;/ui&gt; | 
| Mean Square Error (MSE)   | &lt;ul&gt;&lt;li&gt; regression &lt;/li&gt; &lt;li&gt;difference between actual vs. observed squared&lt;/li&gt;&lt;/ui&gt;         |
| Accuracy   |    &lt;ul&gt;&lt;li&gt; classification &lt;/li&gt; &lt;li&gt;proportion of correctly classified &lt;/li&gt;&lt;/ui&gt;      | 
| F1 Score   | &lt;ul&gt;&lt;li&gt; classification &lt;/li&gt; &lt;li&gt;combines precision and recall&lt;/li&gt;&lt;/ui&gt;        | 

[See formulas here](https://en.wikipedia.org/wiki/Receiver_operating_characteristic)

[J Natl Cancer Inst. 2015 Jun 24;107(8):djv153.](https://pubmed.ncbi.nlm.nih.gov/26109105/)

---

## Apply biomarker panels to test cohorts

.pull-left[

### Diff_plsda_biomarker_panel

#### Train (model built using this data)


```
##        
##         Basal Her2 LumA
##   Basal    43    0    0
##   Her2      2   29    3
##   LumA      0    1   72
```

Accuracy: 0.96 and Balanced Accuracy: 0.96


#### Test (apply model to independent cohort)


```
##        
##         Basal Her2 LumA
##   Basal    18    0    0
##   Her2      3   13    1
##   LumA      0    1   34
```

Accuracy: 0.93 and Balanced Accuracy: 0.92

]

.pull-right[

### splsda_biomarker_panel

#### Train (model built using this data)


```
##        
##         Basal Her2 LumA
##   Basal    43    0    0
##   Her2      2   29    3
##   LumA      0    1   72
```

Accuracy: 0.96 and Balanced Accuracy: 0.96


#### Test (apply model to independent cohort)


```
##        
##         Basal Her2 LumA
##   Basal    20    0    0
##   Her2      1   14    1
##   LumA      0    0   34
```

Accuracy: 0.97 and Balanced Accuracy: 0.97


]

---

## Q: What if you don't have a test cohort?

### Anwers: split data and pretend you do have test cohorts:

#### Resampling strategies

| Type      | Description
| ----------- | ----------- |
| K-fold cross validation  (CV)    |  &lt;ul&gt;&lt;li&gt; split data into K folds of equal size (can also be stratified across disease groups) &lt;/li&gt;&lt;/ul&gt;   
| repeated K-fold cross-validation   |  &lt;ul&gt;&lt;li&gt;repeat K-fold CV multiple times until metric values convergences &lt;/li&gt;&lt;/ul&gt;
| nested cross-validation if hyperparameter tuning is required | &lt;ul&gt;&lt;li&gt;repeat K-fold CV with each fold and select the best model to apply to test data &lt;/li&gt;&lt;/ul&gt;

- bias: how are far is the model away from the true model (K=N has the lowest biased but high variance whereas lower values of K have higher bias but lower variance); K of 5 or 10 offer a good compromise.
---

## K-fold cross validation  (CV)

.center[&lt;img src ="img/cv/Slide1.png" alt="cv_slide1", width="100%"/&gt;]

---

## K-fold cross validation  (4-fold CV)

.center[&lt;img src ="img/cv/Slide2.png" alt="cv_slide2", width="100%"/&gt;]

---

## K-fold cross validation  (4-fold CV)

.center[&lt;img src ="img/cv/Slide3.png" alt="cv_slide3", width="100%"/&gt;]

---

## K-fold cross validation  (4-fold CV)
* Note: many approaches can be used to create the final model
.center[&lt;img src ="img/cv/Slide4.png" alt="cv_slide4", width="100%"/&gt;]

---

## K-fold cross validation  (4-fold CV)
* Note: many approaches can be used to create the final model
.center[&lt;img src ="img/cv/Slide5.png" alt="cv_slide5", width="100%"/&gt;]


---

## K-fold cross validation  (4-fold CV)
* Note: many approaches can be used to create the final model
.center[&lt;img src ="img/cv/Slide6.png" alt="cv_slide6", width="100%"/&gt;]


---

## K-fold cross validation  (4-fold CV)
* Note: many approaches can be used to create the final model
.center[&lt;img src ="img/cv/Slide7.png" alt="cv_slide7", width="100%"/&gt;]


---

## K-fold cross validation  (4-fold CV)
* Note: many approaches can be used to create the final model
.center[&lt;img src ="img/cv/Slide8.png" alt="cv_slide8", width="100%"/&gt;]


---

## K-fold cross validation  (4-fold CV)
* Note: many approaches can be used to create the final model
.center[&lt;img src ="img/cv/Slide9.png" alt="cv_slide9", width="100%"/&gt;]

---

## K-fold cross validation  (4-fold CV)
* Note: many approaches can be used to create the final model
.center[&lt;img src ="img/cv/Slide10.png" alt="cv_slide10", width="100%"/&gt;]

---

## K-fold cross validation  (4-fold CV)
* Note: many approaches can be used to create the final model
.center[&lt;img src ="img/cv/Slide11.png" alt="cv_slide11", width="100%"/&gt;]

---

## Simulation scenerio (true error rate = 50%)

.pull-left[

```r
n &lt;- 100; p &lt;- 1000
group &lt;- rep(c("group1", "group2"), each = n/2)
eset &lt;- matrix(rnorm(n * p), nr = n)
colnames(eset) &lt;- paste0("gene", 1:p)

pvals &lt;- apply(eset, 2, function(i){
  t.test(i~group)$p.value
})

biomarkers &lt;- pvals[order(pvals)][1:10]
as_tibble(data.frame(names = names(biomarkers),
           pvalues = biomarkers))
```

```
## # A tibble: 10 × 2
##    names    pvalues
##    &lt;chr&gt;      &lt;dbl&gt;
##  1 gene223 0.000332
##  2 gene488 0.00204 
##  3 gene93  0.00207 
##  4 gene400 0.00358 
##  5 gene318 0.00378 
##  6 gene288 0.00714 
##  7 gene150 0.00853 
##  8 gene628 0.00943 
##  9 gene557 0.00972 
## 10 gene408 0.0111
```
]

.pull-right[

```r
hist(pvals)
```

&lt;img src="slides_files/figure-html/unnamed-chunk-13-1.png" width="90%" /&gt;

]

---

# How to cross-validate?

.pull-left[
## Wrong way to do cross-validation
 - subset dataset to selected biomarkers than apply cross-validation
 

```r
wrong_way_model &lt;- plsda(X = eset[, names(biomarkers)], Y = group, ncomp=1)

cv &lt;- perf(wrong_way_model, validation = "Mfold", folds = 5, auc=TRUE)
cv$auc
```

```
## $comp1
## AUC.mean   AUC.sd 
##    0.912       NA
```

]

.pull-right[

## Right-way to do cross-validation
- DO NOT subset dataset and repeat the same model building process during each cross-validation train/test fold.


```r
right_way_model &lt;- splsda(X = eset, keepX = 10, Y = group, ncomp=1)

cv &lt;- perf(right_way_model, validation = "Mfold", folds = 5, auc=TRUE)
cv$auc
```

```
## $comp1
## AUC.mean   AUC.sd 
##   0.4956       NA
```

]

---

# I want to include differential expression in my approach
- within each cross-validation fold, start with **ALL** the data and perform differential expression, then use the top features to build the model and apply to the held out/test set.

.pull-left[

```r
cvIndex &lt;- caret::createFolds(factor(group), k = 5)  ## dont load the caret package as it requires the pls packages which clashes with the mixOmics pls()

predictions &lt;- lapply(cvIndex, function(fold){
  ## Create train and test dataests
  Xtrain &lt;- eset[-fold, ]
  Xtest &lt;- eset[fold, ]
  ytrain &lt;- group[-fold]
  
  ## differential expression
  pvals &lt;- apply(Xtrain, 2, function(i){
    t.test(i~ytrain)$p.value
  })
  biomarkers &lt;- pvals[order(pvals)][1:10]
  
  ## plsda model
  model &lt;- mixOmics::plsda(X = Xtrain[, names(biomarkers)], Y = ytrain, ncomp=1)
  as.numeric(predict(model, Xtest[, names(biomarkers)])$predict[, , "dim1"][, "group2"])
})
```
]

.pull-right[

```r
pROC::auc(group[unlist(cvIndex)], unlist(predictions), plot = TRUE)
```

```
## Setting levels: control = group1, case = group2
```

```
## Setting direction: controls &lt; cases
```

&lt;img src="slides_files/figure-html/unnamed-chunk-17-1.png" width="60%" /&gt;

```
## Area under the curve: 0.5632
```
]

---

## Nested cross-validation
 - apply cross-validation within each cross-validation fold (*e.g.* hyperparamter tuning)
 
.center[&lt;img src ="img/ncv.png" alt="Nested Cross-validation", width="60%"/&gt;]
[Bioinformatics
. 2020 May 1;36(10):3093-3098](https://pubmed.ncbi.nlm.nih.gov/31985777/)

---

## Hyperparameter tuning (tune the number of variables to select)

### use a grid of values to find the biomarker panel with the least error rate

.pull-left[

```r
set.seed(1)
test.keepX = c(5, 10, 15, 50, 100, 200)
tune= tune(method = "splsda", 
           X = breast.TCGA$data.train$mrna, 
           Y = breast.TCGA$data.train$subtype, 
           ncomp=1,
           test.keepX = test.keepX, 
           folds=5, dist="centroids.dist", progressBar = FALSE)
```

```
## Calling 'tune.splsda'
```

```r
final_model &lt;- splsda(X = breast.TCGA$data.train$mrna, 
           Y = breast.TCGA$data.train$subtype, 
           ncomp=1,
           keepX = test.keepX[which.min(tune$error.rate)])
selectVar(final_model, ncomp=1)$name
```

```
##  [1] "ZNF552"  "KDM4B"   "PREX1"   "LRIG1"   "CCNA2"   "TTC39A"  "C4orf34" "FUT8"    "ASPM"   
## [10] "MEX3A"   "SLC43A3" "SEMA3C"  "STC2"    "E2F1"    "LMO4"    "MED13L"  "FMNL2"   "DTWD2"  
## [19] "SLC19A2" "FAM63A"  "KIF13B"  "MTL5"    "NTN4"    "CSRP2"   "EPHB3"   "NCAPG2"  "SLC5A6" 
## [28] "MEGF9"   "ELP2"    "ALDH6A1" "HTRA1"   "HN1"     "RTN2"    "SHROOM3" "SIGIRR"  "LIMD2"  
## [37] "IFITM2"  "SNED1"   "SNORA8"  "FRMD6"   "PLCD4"   "MAP3K1"  "ASF1B"   "NME3"    "RTKN2"  
## [46] "ZEB1"    "PBX1"    "ALCAM"   "LAPTM4B" "C18orf1"
```

&gt; the optimal biomarker panel has 50 genes.

]

.pull-right[

```r
plot(tune$error.rate ~ test.keepX, type="l", log="x", ylab = "Error rate")
```

&lt;img src="slides_files/figure-html/unnamed-chunk-19-1.png" width="80%" /&gt;


]

---

## The correct way of determining performance.


```r
right_model &lt;- plsda(X = eset, Y = group, ncomp=1)

cv &lt;- perf(right_model, validation = "Mfold", folds = 5, nrepeat = 5, auc=TRUE)
cv$auc
```

```
## $comp1
##   AUC.mean     AUC.sd 
## 0.53104000 0.04711484
```


---


### Stability of variables in the model

* frequency of variable across folds.

---

### Esimate uncertainty of model parameter estimates

* bootstrap

---

# Should you biomarker? yes

.center[&lt;img src ="img/biomarker_process.png" alt="Biomarker process", width="90%"/&gt;]


---

# Resources

## Models
 - [Tidymodels](https://www.tidymodels.org/)

## Slides
 - [xaringan](https://arm.rbind.io/slides/xaringan.html#137)

## Git
 - [Get Git!](https://uidaholib.github.io/get-git/6resources.html)
 - [Let's Git started](https://happygitwithr.com/)
 - [.gitignore](https://zellwk.com/blog/gitignore/)
 - [GitHub Actions]()

## Cross-validation
- [Resamplign techniques](https://hastie.su.domains/MOOC-Slides/cv_boot.pdf)
 
## Problem Set

1) Git-LFS
2) GH-Pages
3) use RStudio with Git(Hub)
4) advanced git commands: stash, rebase, 

## Solutions
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
