---
title: "EOPE EWAS Visualization"
author: "William Casazza & Amy Inkster"
date: "July 29, 2022"
output: github_document
---

# Running an Epigenome Wide Association Study on Data from GEO

## Introduction

Welcome all to the first in a three part series on working with publicly available data from the 450K array! This worksheet will walk you through the oddities of downloading data from the Gene Expression Omnibus (GEO), all the way to running some linear models to detect sites that are differentially methylated between case and control groups!

### The data

We will be repeating an analysis using processed data from Illumina's 450k DNA methylation array. The samples were taken from placenta, where subjects had different complications during birth, including early and late onset pre-eclampsia (EOPE and LOPE), as well as intrauterine growth restriction (IUGR) . For more information on this study please refer to [Wilson et al, 2016](https://doi-org.ezproxy.library.ubc.ca/10.1093/hmg/ddx391).

### R packages and set-up

The following packages will be required for this tutorial, if you have not already done so you, and you are working on your local machine, you can install them by pasting the following commands into your R console:

If you started R using a singularity image discussed by Will or Amy, these should already be installed for you!

``` r
install.packages("here")
install.packages("BiocManager")
BiocManager::install(c(
  "minfi",  # For analyzing and visualizing DNAme array data
  "GEOquery", # For loading data from GEO
  "Biobase", # For changing the format of data from GEO
  "limma", # For running differential methylation analysis
  "ggplot2" # For making nice plots,
  "reshape2" # For changing data formats
))
```

Once these packages have been installed, you can run the code chunk below to load them for your current R session:

```{r setup,echo=FALSE}
library(here)
library(minfi)
library(lumi)
library(GEOquery)
library(Biobase)
library(limma)
library(ggplot2)
library(reshape2)
i_am("eope_ewas_workflow.Rmd")
```

## Step 1: Downloading and formatting the data

The datasets requried for this tutorial can be downloaded [here](https://ftp.ncbi.nlm.nih.gov/geo/series/GSE100nnn/GSE100197/suppl/). Place this `GSE100197_series_matrix.txt.gz` file and the `GSE100197_RAW.tar` file in the same folder as `eope_ewas_workflow.Rmd`, and proceed by running code in the `eope_ewas_workflow.Rmd` file.

Data from GEO is stored as a special `series_matrix.txt.gz` file. This file typically contains processed count or methylation data, *in addition to* phenotypic information like disease status, sex, and age.[^1].

[^1]: If you are going to work with *raw* data from a methylation array (in the form of an `.idat` file), please look further into the package `minfi`. A nice tutorial on processing this type of data by the maintainer of `minfi` is available [here](https://www.bioconductor.org/help/course-materials/2015/BioC2015/methylation450k.html).

The `getGEO` function loads in this series matrix data as a special *object* and we can access phenotypes and methylation tables from this dataset using the `pData` and `exprs` function respectively.

In addition, we will use the `.tar` file from GEO, which typically contains raw DNA methylation data in the form of IDAT files. We have already parsed this data for you using the `minfi` package into a special object called an `rgset`.

We will also load in an annotation of where the probes on the 450k array are in the genome, that has extra information about them too, which we will use a bit later.

### Loading the raw data from GEO:

```{r load_rgsets}
# the following lines of code untar and unzip idat files into a directory called
# `GSE100197_RAW`, parse the idats into an `rgset` and save it as an .rds object

# untar(here("data/GSE100197_RAW.tar"), exdir = here("data/GSE100197_RAW/"))
# head(list.files(here("data/GSE100197_RAW/"), pattern = "idat"))
# idats <- list.files(here("data/GSE100197_RAW/"), pattern = "idat.gz$", full = TRUE)
# sapply(idats, gunzip, overwrite = TRUE)
# rgset <- read.metharray.exp(here("data/GSE100197_RAW/"))
# colnames(rgset) <- sub("_.*", "", colnames(rgset))
# saveRDS(rgset, here("data/GSE100197_rgset.rds"))

rgset <- readRDS(here("data/GSE100197_rgset.rds"))
```

### Loading the GEO series matrix object:

```{r load_data}
#eset <- getGEO(file="GSE100197_series_matrix.txt.gz")
# alternatively uncomment and run the command below:
# eset <- getGEO("GSE100197"); eset <- eset$GSE100197_series_matrix.txt.gz
eset <- readRDS(here("data/GSE100197.rds"))
eset
```

From the above printout, we can see that the object contains an "assay" stored under `assayData` and variables stored under `phenoData`. These functions can be used for neatly viewing the data, but for our purposes we will need to get them as a `matrix`.

### Load probe annotation

```{r load_anno}
anno <- read.delim(here("data/HM450.hg19.manifest.tsv"))
```

### Formatting the metadata (all data that *isn't* DNA methylation)

We can use `pData` to get the phenotypes of our subjects as a matrix:

```{r format_data}
metadata <- pData(eset)
head(metadata)

# Rename weird columns
colnames(metadata) <- gsub(":ch1| ","",colnames(metadata))
head(metadata)
# Select just the variables we want to analyse
metadata <- metadata[,c("pathologygroup","fetalsex","gestationalage")]
head(metadata)
```

We can now do the following:

1.  Remove Replicate samples

2.  Make sure age is encoded as a NUMBER

```{r clean_data}
# Step 1
metadata <- metadata[!metadata$pathologygroup == "REPLICATE",] 

# Step 2
metadata$gestationalage <- as.numeric(metadata$gestationalage) 

head(metadata)
```

## Step 2: Process DNA methylation data

### Normalization

In this step we will normalize the DNAme data. If you want to learn more about the normalization method we are using, check out the associated [publication](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5408810/).

This step takes a while!

```{r normalize_data}
# normalize, creates a MethylSet class of object with meth/unmeth signal intensities
mset <- preprocessNoob(rgset)

# use normalized mset to calculate normalized beta values
betas <- getBeta(mset)
```

### Detection p value

Illumina calculates a detection p value for each of its probes, which refers to how reliably that probe's fluorescence (which is later used to calculate DNAme) can be distinguished from background fluorescence. We want to restrict our analysis to only probes that are fluorescing significantly differently than the background, so that we can trust their methylated or unmethylated signal (i.e. we want to keep probes that have a low detection p value, or detp \< 0.01). In this step we are going to identify which probes have a high detection p value in \> 10% of our samples (you can change this threshold to be more stringent (5%) or less (20%)), our smallest group size is IUGR with 11 samples, which is roughly 10% of our data, which is why we'll using 10%.

1.  Retrieve the detection p values at all probes in all samples

2.  Calculate how many samples fail per probe

3.  Index probes/rows with \>10% samples failing and remove

```{r detp_filter}
# Step 1
detp <- detectionP(rgset)
detp[1:3,1:3]

# Step 2
detp_fail <- detp > 0.01
detp_fail[1:3,1:3]
detp_fail <- rowSums(detp_fail) > nrow(metadata)*0.1
table(detp_fail)

# step 3 - remove failed probes from betas object
dim(betas)
betas <- betas[!detp_fail,]
dim(betas)
```

### Cross-hybridizing & polymorphic

Some probes on the array were found to have nucleotide sequences that map equally well to multiple regions of the genome. This is a problem because we want to be sure that we know with certainty which genomic locus we are measuring DNAme from. In this case we will therefore remove probes that have been computationally predicted to be non-specific or "cross-hybridizing".

We will also remove probes that sit near single nucleotide polymorphisms in the genome. Genetic variation at these polymorphic sites can affect how well a probe sequence binds to a region, and can therefore interfere with how well the DNAme is measured; SNPs in the last 5bp of a probe with high minor allele frequency have been found to affect DNAme the most. These are typically called "polymorphic" probes, and we will remove them as well.

Zhou et al. (2017) published a comprehensive annotation of these poorly-performing probes, and using their "MASK_general" column, we will mask (remove) probes that demonstrated cross-hybridization potential, or overlapped \>1% minor allele frequency polymorphisms in the last 5bp of the 50-nucleotide probe sequence, closest to the CpG.

More on this annotation is available [here](https://pubmed.ncbi.nlm.nih.gov/27924034/). The annotation itself can be downloaded from the [Zhou lab repository](https://zwdzwd.github.io/InfiniumAnnotation), which also houses their annotations for other array platforms such as the EPIC array.

```{r mask_filter}
# pull probes that are in the MASK_general column
table(anno$MASK_general)
mask <- anno[anno$MASK_general == TRUE, ]

# remove these from the betas object
dim(betas)
betas <- betas[!(rownames(betas) %in% mask$probeID), ]
dim(betas)
```

### Re-ordering our methylation data

Now that we've removed some samples from the metadata objectbased on our phenotype data, and we have normalized and processed our methylation data, we should use the `match` function to match the samples in our `metadata` matrix to those in our methylation `betas` matrix:

```{r reorder}
betas[1:5,1:5]
dim(betas)

matched <- match(rownames(metadata),colnames(betas))
betas <- betas[,matched]
dim(betas)

all(colnames(betas) == rownames(metadata)) # should be TRUE
```

## Step 3: Data visualization

Visualizing data can be a good way to make sure that we 1) understand the format of our data exactly and 2) haven't made any unexpected mistakes in processing.

For example, we can visualize DNA methylation for several samples to make sure that the distribution of $\beta$ values is what we expect. In placenta, we expect for most CpG sites that DNA methylation will be low or high. However, an interesting quirk of placental DNA methylation is that many sites will show intermediate levels of methylation as well, giving our density distribution a characteristic "three-hump" shape.

Let's plot the distribution of all $\beta$ values for the first 5 subjects:

```{r data_viz}
betas[1:5,1:5]

# first reformat data for ggplot
to_plot <- melt(
  betas[,1:5],
  value.name = "beta",
  varnames = c("Probes","Subjects") # c(rownames,colnames)
)
head(to_plot)
ggplot(to_plot,aes(x=beta,color=Subjects)) +
  geom_density()
```

Lets also look at the phenotypic characteristics of our data:

```{r pheno}
ggplot(metadata,aes(pathologygroup)) + geom_bar()

ggplot(metadata,aes(x=pathologygroup, fill=fetalsex)) + 
  geom_bar() 
ggplot(metadata,aes(x=pathologygroup, fill=fetalsex)) + 
  geom_bar(position="fill")
ggplot(metadata,aes(x=pathologygroup, fill=fetalsex)) + 
  geom_bar(position="dodge")
ggplot(metadata,aes(x=pathologygroup, y=gestationalage)) + 
  geom_boxplot() 
```

## Step 4: Fit linear model and relevant contrasts

Now it's time to test whether there is differential DNA methylation between groups. Using *linear modeling*, we can examine this relationship at each CpG.

You might remember learning the formula *y = mx + b* in math class, where "m" represents the slope, and "b" represents the intercept. We can model the relationship between CpG methylation and our variable of interest in the same way.

We consider "y" to be the level of DNA methylation, or beta value, and we consider "x" to be our variable of interest, in this case the pathology group.

![Linear Regression](https://www.dataquest.io/wp-content/uploads/2019/12/linear-modeling-in-r-course.jpg)

However, each CpG site might have a different relationship between methylation and pathology group. For example, methylation levels at CpGs located in genes related to fetal growth might have a stronger relationship with pathology (a larger "m" value, or slope), whereas methylation levels at CpGs located outside of genes or in genes with totally unrelated functions could have no relationship with pathology (m=0).

![Possible X-Y relationships in linear regression](https://3.bp.blogspot.com/-s7T8XRynMKI/VT43mN1qY5I/AAAAAAAAAcs/lgrgkOxB2wo/s1600/Screen%252BShot%252B2015-04-27%252Bat%252B9.19.53%252BAM.png)

In differential DNA methylation analysis, we will *run a linear model across each CpG* to evaluate its relationship with with pathology. To accomplish this, we will use the R package *limma* (Smyth et al., 2002). Limma will fit the model across every CpG, and return statistics including the coeffificent (slope) for the association between DNAm and pathology, and p-values indiciating whether the association is statistically significant. Then, we can narrow down the list to our "hits", or only the CpG sites which are differentially methylated (p \< 0.05 for association between DNAm and pathology).

### Designing your model

In this study, there are multiple pathology groups, and data on other variables such as fetal sex and gestational age. It's important to *define your research question* before designing your model, so that we can compare the right groups to each other.

For this example, we hypothesized that pathology group and fetal sex will each affect placental DNAm patterns. We can look at the effects of pathology group and fetal sex using one formula, with the function *model.matrix()* in limma.

```{r model}
model <- model.matrix(~ 0+ pathologygroup + fetalsex, data=metadata)
head(model)
```

The object "model" consists of a table detailing which samples belong to which pathology group as well as the fetal sex. Since there are only two sexes, there is one column for whether the fetal sex is male (yes/no). Since there are five pathology groups, each group has its own column, with a yes/no indiciating whether the sample is part of that group.

Feeding this table into the *lmFit()* function, along with an object containing our DNAm data, will ask limma to examine the relationship between DNA methylation and each of the five pathology groups one by one, as well as the relationship between DNA methylation and fetal sex. *eBayes()* will derive statistics such as p-values from this fit.

Note: before we do this, we are going to convert our beta values to a metric called M-values, which is a logit-transformed beta value that better fits the assumptions of linear modelling. More on that [here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3012676/) if you're interested.

```{r fit}
# convert to m values, plot distribution
mvals <- beta2m(betas)

to_plot_m <- melt(
  mvals[,1:5],
  value.name = "M",
  varnames = c("Probes","Subjects") # c(rownames,colnames)
)
head(to_plot_m)
ggplot(to_plot_m,aes(x=M,color=Subjects)) +
  geom_density()

# now fit the linear model
fit <- lmFit(mvals, model)
fit <- eBayes(fit)
```

### Extracting comparisons of interest

What if we want to compare DNAm between two different pathology groups instead of looking at them individually? To do this, we create a *contrast matrix*, which is a table that tells limma which two groups to compare to one another.

Here, we will contrast placentas from mothers with preterm infants against placentas from mothers with early onset pre-eclampsia, late-onset pre-eclampsia, and intrauterine growth restriction. We'll do the same contrasts for term infants. We will leave sex alone, to look at the effect of sex across all the groups at the same time.

```{r contrast}
contrasts <- makeContrasts(
  preTEOPE=pathologygroupPreT-pathologygroupEOPE,
  # preTLOPE=pathologygroupPreT-pathologygroupLOPE,
  # preTIUGR=pathologygroupPreT-pathologygroupIUGR,
  # termEOPE=pathologygroupTerm-pathologygroupEOPE,
  termLOPE=pathologygroupTerm-pathologygroupLOPE,
  termIUGR=pathologygroupTerm-pathologygroupIUGR,
  sex=fetalsexMALE,levels=model)
head(contrasts)
```

Each comparison is indicated in a column of the contrast matrix, with "1" indicating the reference group, and "-1" indicating the group we are comparing against the reference. For example, the column "preTEOPE" compares preterm placentas against early onset pre-eclampsia placentas, where preterm placentas are the reference group. Looking at the rows, the "pathologygroupPreT" has a value of "1", whereas the "pathologygroupEOPE" has a value of "-1". The "0" in the other rows indicates those groups are not part of this comparison.

Next, we use the function *contrasts.fit()* to fit our methylation data to these comparisons, and run *eBayes()* again to derive statistics:

```{r cont_fit}
fitCont <- contrasts.fit(fit,contrasts)
fitCont <- eBayes(fitCont)
```

## Extract stats from relevant comparisons

Let's take a closer look at the results from two of the above contrasts: "preTEOPE" and "sex". This is equivalent to asking the following research questions:

1\) *Are there DNAm differences between placentas from mothers with preterm infants (PreT) and mothers with early onset pre-eclampsia (EOPE)?*

2\) *Are there DNAm differences between placentas from mothers who gave birth to male versus female infants?*

The function *topTable()* lets us extract detailed statistics information for a single contrast of interest. We can set cutoffs for significance and effect size such that only the information for significant CpG sites is returned. Here, we will use an adjusted p-value cutoff of \<0.05 and a log fold change cutoff of \>0.1.

```{r extract_stats}
eope_stats <- topTable(fitCont,coef = "preTEOPE", adjust.method = "BH", p.value = 0.05, lfc = 0.1, number = Inf)
nrow(eope_stats)
head(eope_stats)
eope_sex <- topTable(fitCont,coef = "sex", adjust.method = "BH", p.value = 0.05, lfc = 0.1, number = Inf)
nrow(eope_sex)
head(eope_sex)
```

## Step 5: Plotting results from our EWAS

### Volcano plot

'Omics association study results are typically visualized using a *volcano plot*, where the effect size is on the x axis and the -log p-value is on the y-axis. This makes a "volcano" shape such that significant hits look like they are "erupting" out into the top corners of the plot.

The below code also categorizes hits by if they are significant with increased methylation (DME_pos), significant with decreased methylation (DME_neg), or non-significant (not DME). This allows us to colour these points by category for easier visualization of hits.

```{r volcano}
plot_stats <- function(cont){
  all_eope_stats <- topTable(fitCont,coef = cont, adjust.method = "BH",number = Inf)
  all_eope_stats$color <- ifelse(
    all_eope_stats$adj.P.Val <0.05,
    ifelse(
      all_eope_stats$logFC > 0.1,
      "DME_pos",
      ifelse(all_eope_stats$logFC < -0.1,
        "DME_neg",
        "not DME"
      )),
    "not DME")
  p <- ggplot(all_eope_stats, aes(logFC,-log10(adj.P.Val),color=color)) + 
    geom_point(size=1,alpha=0.3)+
    scale_color_manual(values = c("DME_pos"="dark green","DME_neg"="red","not DME"="black"))+
    coord_cartesian(xlim=c(-0.3,0.3),ylim=c(0,10))
  print(p)
  return(all_eope_stats)
}
eope_all_stat <-plot_stats("preTEOPE")
lope_all_stat<- plot_stats("sex")
```
