## UBC's Precision Health Bootcamp 2-part RNA-seq Workshop 2021 
<img src= "https://user-images.githubusercontent.com/59856969/181839760-0bda1af3-3814-4df6-8416-da099a6eff16.png" width = 150/>

<br> 

### Requirements: 

* Basic working-knowledge and experience with R  
* Complete the [Pre-worshop To-Do's](https://github.com/Phillip-a-richmond/PrecisionHealthVirtualEnvironment/blob/main/Workshops/Applied_RNA-seq/Pre_Workshop_ToDos.md) 

<br>

**Part 1 - RNA-seq: From Raw to Processed Data**  
 
Part 1 of the workshop will introduce what RNA-seq data looks like and how to clean the data. Starting with a raw expression count matrix (pre-formatted for participantsâ€™ convenience), we will go through steps to visualize, filter, and normalize the data. At the end of Part 1, our processed data will be ready to perform differential gene analysis and gene-pathway enrichment on.
 
*Note: While this workshop will not focus on the upstream quality control of FASTQ files and genome alignment, we will briefly go through how these steps are performed. The reason for not including the above mentioned processes in the workshop is due to the need of a linux computing system required for alignment, which some of the audience might not have immediate access to.*  

<br>

**Part 2 - RNA-seq: From Differential Expression and Beyond**  
 
Part 2 of the workshop will start with processed counts data and run through how to conduct principle component analysis, differential expression, as well as introduce tools for gene and pathway analysis.

<br>  

***  

Both workshops are stand-alone, wherein Part 1 is focused on QC and Part 2 focuses on downstream analysis; however, we do encourage to attend both. Both workshops will be interactive with small, real-time exercises present throughout, as our goal is to deliver a well-rounded walkthrough of handling and processing RNA-seq data.

